platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/problems/ConfigurationCacheProblems.kt

```diff
@@ -171,31 +181,35 @@ class ConfigurationCacheProblems(
     }
 
     override fun forIncompatibleTask(trace: PropertyTrace, reason: String): ProblemsListener {
+        onIncompatibleTask(trace, reason)
+        return ErrorsAreProblemsProblemsListener(failureFactory, ProblemSeverity.Suppressed)
+    }
+
+    override fun forTask(task: Task): ProblemsListener {
+        val degradationReasons = degradationController.degradationReasons[task]
+        return if (!degradationReasons.isNullOrEmpty()) {
+            onIncompatibleTask(locationForTask(task), degradationReasons.joinToString())
+            ErrorsAreProblemsProblemsListener(failureFactory, ProblemSeverity.SuppressedSilently)
+        } else this
+    }
+
+    fun shouldDegradeGracefully(): Boolean {
+        degradationController.collectDegradationReasons()
+        return degradationController.degradationReasons.isNotEmpty()
```

Comment:
:x: This is a bad design. What if some other code reads `degradationReasons` without calling `collectDegradationReasons` first?

If you don't want to make the computation lazy (though I don't see any particular problem with that), let's have a state check at least.

Reply:
I agree with @mlopatkin here